src/main/java/com/google/sps/servlets/SearchServlet.java

```
@@ -55,46 +55,121 @@
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
-    throws ServletException, IOException {
-    String stringToMatch = request
-      .getParameter("searchterm")
-      .replaceAll("[^a-zA-Z0-9]+", " ")
-      .trim();
-    request.setAttribute("dataSearched", stringToMatch);
+    throws ServletException, IOException, NullPointerException {
+    List<Recipe> recipesListToReturn = new ArrayList<>();
+    Integer timeValue;
+    String stringToMatch = getParameter("searchterm", request);
+    String difficulty = getParameter("difficulty", request);
+    String time = getParameter("time", request);
+    HashMap<String, String> difficultiesMap = createMapOfDifficulties();
+    HashMap<Integer, String> timeIntervalsMap = createMapOfTimeIntervals();
 
-    List<Recipe> recipesListToReturn = recipesMatching(
-      request,
-      response,
-      stringToMatch
+    request.setAttribute("dataSearched", stringToMatch);
+    request.setAttribute("difficultySearchedKey", difficulty);
+    request.setAttribute("timeSearchedKey", time);
+    request.setAttribute(
+      "difficultySearchedValue",
+      difficultiesMap.get(difficulty)
     );
+    request.setAttribute("difficultyList", difficultiesMap);
+    request.setAttribute("cookingTimeList", timeIntervalsMap);
 
+    stringToMatch = trimAndRemoveCommas(stringToMatch);
+    difficulty = trimAndRemoveCommas(difficulty);
+    try {
+      timeValue = Integer.valueOf(trimAndRemoveCommas(time));
+    } catch (NumberFormatException e) {
+      timeValue = 0;
+    }
+    request.setAttribute("timeSearchedValue", timeIntervalsMap.get(timeValue));
+
+    try {
+      recipesListToReturn =
+        recipesMatching(
+          request,
+          response,
+          stringToMatch,
+          difficulty,
+          timeValue
+        );
+    } catch (SearchQueryException e) {
+      request.getRequestDispatcher("/search.jsp").forward(request, response);
```

You need to return an error code in this case.